### ISSUE TITLE
Flaky tests

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#5390]
Trying to collect flaky tests here, in the hope of getting the CI to a "normally green" state again.

I've fixed the usual suspects over the last couple of days (e.g. d8d9311c2e1fe9c3981febe15774aec8849a2934 + a4abf0f535c1314048ada8a5e3fef951c0889a27, cbc835acb721840236f1abd5006e11ebb1810495, 5985353c4a8d9846fb4b198b19c9210962faf593, 513c8343ca6c89ec4bd2454c5b8b0fce0fe6ee36) but there might still be some remaining. I'll try to keep an eye on failing Travis/Appveyor tests, but probably not for WIP PRs - so if you see something fail which isn't your code's fault, please put a link here.

**Comment by The-Compiler** [comment]
# test_single_message[MessageLevel.error]

Especially on macOS? See https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/680022055

```
tests/unit/mainwindow/test_messageview.py ..F
___________________ test_single_message[MessageLevel.error] ____________________
qtbot = <pytestqt.qtbot.QtBot object at 0x1292d3e10>
view = <qutebrowser.mainwindow.messageview.MessageView object at 0x1235d0550>
level = <MessageLevel.error: 1>
    @pytest.mark.parametrize('level', [usertypes.MessageLevel.info,
                                       usertypes.MessageLevel.warning,
                                       usertypes.MessageLevel.error])
    def test_single_message(qtbot, view, level):
        with qtbot.waitExposed(view, timeout=5000):
>           view.show_message(level, 'test')
E           pytestqt.exceptions.TimeoutError: widget <qutebrowser.mainwindow.messageview.MessageView object at 0x1235d0550> not exposed in 5000 ms.
tests/unit/mainwindow/test_messageview.py:40: TimeoutError
py-3008.log-setup.0
------------------------------ Captured log setup ------------------------------
DEBUG    config:config.py:324 Config option changed: messages.timeout = 100
py-3008.log.0
------------------------------ Captured log call -------------------------------
VDEBUG   config:stylesheet.py:113 stylesheet for Message: 
            padding-top: 2px;
            padding-bottom: 2px;
        
                background-color: red;
                color: white;
                font: 10pt Monaco;
                border-bottom: 1px solid #bb0000;
```

For some reason, 0fec91584dd9b399a9c2e85cc7b18316a2984e43 didn't help.

**Comment by The-Compiler** [comment]
# munmap_chunk(): invalid pointer 

https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/680117274

```
E           end2end.fixtures.testprocess.InvalidLine: 
E           munmap_chunk(): invalid pointer
E           Fatal Python error: Aborted
E           Current thread 0x00007f2353e74740 (most recent call first):
E           <no Python frame>
```

Probably not much we can do about that one.

**Comment by The-Compiler** [comment]
# libdl inconsistency

https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/680133532

Looks like our workaround isn't working properly:

```
E               end2end.fixtures.testprocess.ProcessExited: 
E               NOTE: Restarted after libc DL inconsistency!
```

**Comment by The-Compiler** [comment]
# IPC socket timeout

https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/680133438

```
________________________ test_qute_settings_persistence ________________________
short_tmpdir = local('/tmp/tmp0tjkn890')
request = <FixtureRequest for <Function test_qute_settings_persistence>>
quteproc_new = <end2end.fixtures.quteprocess.QuteProc object at 0x7f14246b3550>
    def test_qute_settings_persistence(short_tmpdir, request, quteproc_new):
        """Make sure settings from qute://settings are persistent."""
        args = _base_args(request.config) + ['--basedir', str(short_tmpdir)]
        quteproc_new.start(args)
        quteproc_new.open_path('qute://settings/')
        quteproc_new.send_cmd(':jseval --world main '
                              'cset("search.ignore_case", "always")')
        quteproc_new.wait_for(message='No output or error')
    
        assert quteproc_new.get_setting('search.ignore_case') == 'always'
    
        quteproc_new.send_cmd(':quit')
        quteproc_new.wait_for_quit()
    
        quteproc_new.start(args)
>       assert quteproc_new.get_setting('search.ignore_case') == 'always'
tests/end2end/test_invocations.py:391: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/end2end/fixtures/quteprocess.py:744: in get_setting
    self.send_cmd(cmd)
tests/end2end/fixtures/quteprocess.py:730: in send_cmd
    self.send_ipc([command])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <end2end.fixtures.quteprocess.QuteProc object at 0x7f14246b3550>
commands = [':set search.ignore_case?'], target_arg = ''
    def send_ipc(self, commands, target_arg=''):
        """Send a raw command to the running IPC socket."""
        delay = self.request.config.getoption('--qute-delay')
        time.sleep(delay / 1000)
    
        assert self._ipc_socket is not None
        ipc.send_to_running_instance(self._ipc_socket, commands, target_arg)
>       self.wait_for(category='ipc', module='ipc', function='on_ready_read',
                      message='Read from socket *')
E       end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'category': 'ipc', 'module': 'ipc', 'function': 'on_ready_read', 'message': 'Read from socket *'}.
```

No idea what's going on there...

**Comment by The-Compiler** [comment]
# Segfault in caret tests

https://github.com/qutebrowser/qutebrowser/issues/5390

```
tests/unit/browser/test_caret.py sERROR: InvocationError for command /home/travis/build/qutebrowser/qutebrowser/.tox/py38-pyqt514-cov/bin/python -bb -m pytest tests (exited with code 127)
```

(exit code 127 seems weird though, that's usually what bash uses when a command wasn't found...)

**Comment by The-Compiler** [comment]
# Hypothesis tests with ListOrValue

https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/682219548

Our condition is supposed to skip this, and I don't understand why it does not!

```
__________________ TestAll.test_from_str_hypothesis[klass19] ___________________
self = <test_configtypes.TestAll object at 0x7f6443305208>
klass = functools.partial(<class 'qutebrowser.config.configtypes.List'>, valtype=<qutebrowser.config.configtypes.Url none_ok=False>)
    @pytest.mark.usefixtures('config_tmpdir')
>   @hypothesis.given(strategies.text())
    @hypothesis.example('\x00')
    def test_from_str_hypothesis(self, klass, s):
tests/unit/config/test_configtypes.py:239: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <test_configtypes.TestAll object at 0x7f6443305208>
klass = functools.partial(<class 'qutebrowser.config.configtypes.List'>, valtype=<qutebrowser.config.configtypes.Url none_ok=False>)
s = '- A'
    @pytest.mark.usefixtures('config_tmpdir')
    @hypothesis.given(strategies.text())
    @hypothesis.example('\x00')
    def test_from_str_hypothesis(self, klass, s):
        typ = klass()
        try:
            val = typ.from_str(s)
        except configexc.ValidationError:
            return
    
        # For some types, we don't actually get the internal (YAML-like) value
        # back from from_str(), so we can't convert it back.
        if klass in [configtypes.FuzzyUrl, configtypes.QtFont,
                     configtypes.ShellCommand, configtypes.Url]:
            return
    
        converted = typ.to_str(val)
        # For those we only check that to_str doesn't crash, but we can't be
        # sure we get the 100% same value back.
        if klass in [
                configtypes.Bool,  # on -> true
                configtypes.BoolAsk,  # ditto
                configtypes.Float,  # 1.0 -> 1
                configtypes.Int,  # 00 -> 0
                configtypes.PercOrInt,  # ditto
        ]:
            return
        elif (isinstance(typ, functools.partial) and
              isinstance(typ.func, configtypes.ListOrValue)):
            # "- /" -> "/"
            return
        elif (isinstance(typ, configtypes.ListOrValue) and
              isinstance(typ.valtype, configtypes.Int)):
            # "00" -> "0"
            return
    
>       assert converted == s
E       assert '["A"]' == '- A'
E         - - A
E         + ["A"]
```

**Comment by The-Compiler** [comment]
# test_followselected_with_link_tabbing on Windows

https://ci.appveyor.com/project/qutebrowser/qutebrowser/builds/32598292/job/nfeqq6ma8kcgvam1

```
E       end2end.fixtures.testprocess.WaitForTimeout: Timed out while waiting for http://localhost:1063/data/hello.txt to be loaded
```

**Comment by The-Compiler** [comment]
# test_inserting_text_into_a_text_field_at_specific_position on Windows

https://ci.appveyor.com/project/qutebrowser/qutebrowser/builds/32680115/job/8iss2ta0cqen9g3i

```
E       end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'category': 'js', 'message': '[*] textarea contents: onHello worlde two three four'}.
```

**Comment by The-Compiler** [comment]
# test_spawning_an_editor_and_saving

https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/684194302

```
E       end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'category': 'js', 'message': '[*] text: foobar'}.
```

**Comment by The-Compiler** [comment]
# test_qute_settings_persistence

https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/686732555

```
________________________ test_qute_settings_persistence ________________________
short_tmpdir = local('/tmp/tmpkgqim7od')
request = <FixtureRequest for <Function test_qute_settings_persistence>>
quteproc_new = <end2end.fixtures.quteprocess.QuteProc object at 0x7ff772d331f0>
    def test_qute_settings_persistence(short_tmpdir, request, quteproc_new):
        """Make sure settings from qute://settings are persistent."""
        args = _base_args(request.config) + ['--basedir', str(short_tmpdir)]
        quteproc_new.start(args)
        quteproc_new.open_path('qute://settings/')
        quteproc_new.send_cmd(':jseval --world main '
                              'cset("search.ignore_case", "always")')
        quteproc_new.wait_for(message='No output or error')
    
>       assert quteproc_new.get_setting('search.ignore_case') == 'always'
E       AssertionError: assert 'smart' == 'always'
E         - always
E         + smart
```

**Comment by The-Compiler** [comment]
# test_open_with_ascii_locale

https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/687381927

```
11:09:13.494 DEBUG    webview    browsertab:_set_load_status:937 load status for <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=0 url='file:///f%C3%B6%C3%B6.html'>: LoadStatus.error
11:09:13.494 DEBUG    signals    signalfilter:_filter_signals:88 emitting: cur_load_status_changed(<LoadStatus.error: 4>) (tab 0)
----> found it
11:09:13.509 ERROR    webview    webenginetab:_error_page_workaround:1526 Load error: ERR_FILE_NOT_FOUND
```

**Comment by The-Compiler** [comment]
# Missing load_finished in caret tests

https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/687381933

```
_ ERROR at setup of TestBlock.test_moving_back_to_the_start_of_prev_block[webengine] _
web_tab = <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=7 url='qute://testdata/data/caret.html'>
qtbot = <pytestqt.qtbot.QtBot object at 0x118695310>
mode_manager = <qutebrowser.keyinput.modeman.ModeManager mode=<KeyMode.normal: 1>>
    @pytest.fixture
    def caret(web_tab, qtbot, mode_manager):
        with qtbot.wait_signal(web_tab.load_finished):
>           web_tab.load_url(QUrl('qute://testdata/data/caret.html'))
E           pytestqt.exceptions.TimeoutError: Signal load_finished(bool) not emitted after 1000 ms
```

**Comment by The-Compiler** [comment]
# TestFileHandling.test_backup

on macOS: https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/688931339

```
tests/unit/misc/test_editor.py .........F
_________________________ TestFileHandling.test_backup _________________________
self = <test_editor.TestFileHandling object at 0x134887610>
qtbot = <pytestqt.qtbot.QtBot object at 0x1348a8390>
message_mock = <helpers.messagemock.MessageMock object at 0x1348a82d0>
    def test_backup(self, qtbot, message_mock):
        editor = editormod.ExternalEditor(watch=True)
        editor.edit('foo')
        with qtbot.wait_signal(editor.file_updated):
>           _update_file(editor._filename, 'bar')
E           pytestqt.exceptions.TimeoutError: Signal file_updated(QString) not emitted after 1000 ms
```

**Comment by The-Compiler** [comment]
# Double free or corruption

https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/691232956

```
10:04:49.655 DEBUG    destroy    objreg:on_destroyed:125 schedule removal: download-model
INVALID: double free or corruption (out)
INVALID: Fatal Python error: Aborted
IGNORED: 
INVALID: Current thread 0x00007f24a159e740 (most recent call first):
INVALID: <no Python frame>
10:05:05.671 DEBUG    init       earlyinit:init_log:276 Log initialized.
```

**Comment by The-Compiler** [comment]
# Failed to report HostNotFoundError

on macOS:

- https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/691301855

```
______________ TestProxyFromUrl.test_proxy_from_url_pac[pac+http] ______________
tests/unit/utils/test_urlutils.py:719: Failure: Qt messages with level WARNING or above emitted
----------------------------- Captured Qt messages -----------------------------
    QtWarningMsg: QHttpNetworkConnectionPrivate::_q_hostLookupFinished could not de-queue request, failed to report HostNotFoundError
```

I guess that warning could just be ignored?

**Comment by The-Compiler** [comment]
# RuntimeErrors during test_edit_a_command_and_run_it cleanup

https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/701738792

```
18:44:50.634 DEBUG    destroy    quitter:_shutdown_3:270 Now calling QApplication::exit.
18:44:50.634 ERROR    misc       crashsignal:_handle_early_exits:245 Uncaught exception
Traceback (most recent call last):
  File "/home/travis/build/qutebrowser/qutebrowser/qutebrowser/browser/commands.py", line 1421, in <lambda>
    ed.editing_finished.connect(lambda: mainwindow.raise_window(
  File "/home/travis/build/qutebrowser/qutebrowser/qutebrowser/mainwindow/mainwindow.py", line 107, in raise_window
    window.activateWindow()
RuntimeError: wrapped C/C++ object of type MainWindow has been deleted
18:44:50.635 ERROR    destroy    crashsignal:_get_exception_info:215 Error while recovering pages: wrapped C/C++ object of type TabWidget has been deleted
Traceback (most recent call last):
  File "/home/travis/build/qutebrowser/qutebrowser/qutebrowser/misc/crashsignal.py", line 213, in _get_exception_info
    pages = self._recover_pages(forgiving=True)
  File "/home/travis/build/qutebrowser/qutebrowser/qutebrowser/misc/crashsignal.py", line 140, in _recover_pages
    for tab in tabbed_browser.widgets():
  File "/home/travis/build/qutebrowser/qutebrowser/qutebrowser/mainwindow/tabbedbrowser.py", line 288, in widgets
    for i in range(self.widget.count()):
RuntimeError: wrapped C/C++ object of type TabWidget has been deleted
```

**Comment by The-Compiler** [comment]
# Double "element vanished" log in test_spawning_an_editor_and_closing_the_tab

https://github.com/qutebrowser/qutebrowser/runs/879156079?check_suite_focus=true

```
E       Failed: Logged unexpected errors:
E       
E       19:42:02.210 ERROR    message    message:error:59 Edited element vanished
```

```
  19:42:02.190 DEBUG    procs      editor:_on_file_changed:160 Read back: 
  19:42:02.192 ERROR (expected) message    message:error:59 Edited element vanished

  19:42:02.210 DEBUG    procs      editor:_on_file_changed:160 Read back: foobar
  19:42:02.210 ERROR    message    message:error:59 Edited element vanished

  19:42:02.211 INFO     message    message:info:82 Editor backup at /tmp/qutebrowser-editor-backup-nia76073
```

**Comment by The-Compiler** [comment]
# Missing scroll position changed log in test_scrolling_to_right_and_to_left_with_scrolltoperc

https://github.com/qutebrowser/qutebrowser/runs/879155864?check_suite_focus=true

```
  ============================== :scroll-to-perc --horizontal 100 ==============================
  ----> Already found 'Scroll position changed to PyQt5.QtCore.QPoint(*, *)' in the log: 19:46:43.296 VDEBUG   webview    browsertab:_log_scroll_pos_change:557 Scroll position changed to PyQt5.QtCore.QPoint(2495, 0)
```

(but I can't actually find that line in the log?!)

```
  ============================== :scroll-to-perc --horizontal 0 ==============================
  ----> Waiting for 'command called: *' in the log
  19:46:43.619 DEBUG    ipc        ipc:on_error:251 Socket 0x7f4b686ec5f0: error 1: QLocalSocket: Remote closed
  19:46:43.619 DEBUG    ipc        ipc:on_disconnected:293 Client disconnected from socket 0x7f4b686ec5f0.
  19:46:43.619 DEBUG    ipc        ipc:handle_connection:268 No new connection to handle.
  19:46:43.619 DEBUG    init       app:process_pos_args:257 Startup cmd ':scroll-to-perc --horizontal 0'
  19:46:43.619 DEBUG    commands   command:run:542 command called: scroll-to-perc ['--horizontal', '0']
  19:46:43.620 DEBUG    commands   command:run:557 Calling qutebrowser.components.scrollcommands.scroll_to_perc(<qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=25 url='http://localhost:52881/data/scroll/simple.html'>, None, 0.0, True)
  ----> found it
  
  ----> Waiting for 'Scroll position changed to PyQt5.QtCore.QPoint()' in the log
  19:46:48.661 DEBUG    ipc        ipc:on_timeout:386 on_timeout got called with None socket!
  19:46:53.660 DEBUG    ipc        ipc:on_timeout:386 on_timeout got called with None socket!
  19:46:58.663 DEBUG    ipc        ipc:on_timeout:386 on_timeout got called with None socket!
```

```
E       end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'category': 'webview', 'message': 'Scroll position changed to PyQt5.QtCore.QPoint()'}.
```

**Comment by bnavigator** [comment]
#  unreliable test timing in `TestList.test_hypothesis[ListSubclass]`

[qutebrowser_sip5_log.txt](https://github.com/qutebrowser/qutebrowser/files/5358888/qutebrowser_sip5_log.txt)

```
[  271s] =================================== FAILURES ===================================
[  271s] ____________________ TestList.test_hypothesis[ListSubclass] ____________________
[  271s] [gw0] linux -- Python 3.8.5 /usr/bin/python3
[  271s] 
[  271s] self = <test_configtypes.TestList object at 0x7f503875bc40>
[  271s] klass = <class 'test_configtypes.ListSubclass'>
[  271s] 
[  271s]     @hypothesis.given(val=strategies.lists(strategies.just('foo')))
[  271s] >   def test_hypothesis(self, klass, val):
[  271s] 
[  271s] tests/unit/config/test_configtypes.py:665: 
[  271s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[  271s] /usr/lib/python3.8/site-packages/hypothesis/core.py:637: in execute_once
[  271s]     self.__flaky(
[  271s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[  271s] 
[  271s] self = <hypothesis.core.StateForActualGivenExecution object at 0x7f502408f8b0>
[  271s] message = "Hypothesis test_hypothesis(self=<test_configtypes.TestList at 0x7f503875bc40>, klass=ListSubclass, val=['foo', 'foo',... 'foo', 'foo', 'foo', 'foo']) produces unreliable results: Falsified on the first call but did not on a subsequent one"
[  271s] 
[  271s]     def __flaky(self, message):
[  271s]         if len(self.falsifying_examples) <= 1:
[  271s] >           raise Flaky(message)
[  271s] E           hypothesis.errors.Flaky: Hypothesis test_hypothesis(self=<test_configtypes.TestList at 0x7f503875bc40>, klass=ListSubclass, val=['foo', 'foo', 'foo', 'foo', 'foo', 'foo', 'foo']) produces unreliable results: Falsified on the first call but did not on a subsequent one
[  271s] 
[  271s] /usr/lib/python3.8/site-packages/hypothesis/core.py:847: Flaky
[  271s] ---------------------------------- Hypothesis ----------------------------------
[  271s] Falsifying example: test_hypothesis(
[  271s]     self=<test_configtypes.TestList at 0x7f503875bc40>,
[  271s]     klass=test_configtypes.ListSubclass,
[  271s]     val=['foo', 'foo', 'foo', 'foo', 'foo', 'foo', 'foo'],
[  271s] )
[  271s] Unreliable test timings! On an initial run, this test took 1038.29ms, which exceeded the deadline of 600.00ms, but on a subsequent run it took 2.15 ms, which did not. If you expect this sort of variability in your test timings, consider turning deadlines off for this test by setting deadline=None.
```

**Comment by The-Compiler** [comment]
@bnavigator Do you have more details about the environment you did run those tests in? I don't think this is qutebrowser's fault FWIW.

**Comment by bnavigator** [comment]
Hi,

this is from the openSUSE build service: https://build.opensuse.org/package/show/home:bnavigator:sip5/qutebrowser
The linked build log [qutebrowser_sip5_log.txt](https://github.com/qutebrowser/qutebrowser/files/5358888/qutebrowser_sip5_log.txt) should list everything about the build environment.

**Comment by The-Compiler** [comment]
@bnavigator I meant more in terms of hardware - it seems to be a VM (?), is this on a host with many such VMs running in parallel, where it's expected for something that runs relatively quickly to take much more time randomly (>1s in this case)?

Also, you might want to make sure your PyYAML is built with C extensions, e.g. via `python3 -c "from yaml import CLoader"`.

**Comment by bnavigator** [comment]
Yes, these are [VMs](https://build.opensuse.org/monitor) The failure was on lamb24. You can also run the chroot environment on a local machine. Would have to check if the flakyness occurs on those as well.

CLoader can be imported from yaml

**Comment by bnavigator** [comment]
https://github.com/qutebrowser/qutebrowser/issues/5787#issuecomment-709981808

**Comment by toofar** [comment]
What do you think about something like https://github.com/marketplace/buildpulse/ to track flaky tests? We talked on https://github.com/qutebrowser/qutebrowser/pull/7687 about this an I had some ideas about building something to track these ourselves, but I really don't want to :sweat_smile:
From a quick search on the github marketplace I saw https://github.com/marketplace/testspace-com too but that seems more focused on manual testing notes overall.

**Comment by The-Compiler** [comment]
I definitely like the idea of using some sort of tooling for this! I'm a bit lukewarm about installing GitHub Apps for security reasons (see e.g. [what happened with Deepsource](https://discuss.deepsource.com/t/security-incident-on-deepsource-s-github-application/131/2)), but yeah, I also don't see us building something from scratch...

Unfortunately, I don't see Buildpulse working out. Their pricing seems to be:

- Free up to 10k test results per day
- $120/month (!) up to 100k test results per day
- $300/month up to 300k test results per day
- $500/month up to 1M test results per day

I'm assuming that "test results" mean individual tests (otherwise, those numbers likely wouldn't be this high). According to a `tox -e py311-pyqt62 -- --collect-only`, we have 9845 tests in a single pytest run. We currently run 14 different CI jobs, so we're at almost **138k** test results *for a single CI run* (so, realistically, in their $500/month plan...).

Somehow this can't be right, but I can't see how else it'd work. Nevertheless, to give them the benefit of doubt, I'm setting it up and giving it a try. Right now I'm stuck at:

> 🔜 We're provisioning your credentials, and they'll be available here soon! Please check back later today.

**Comment by The-Compiler** [comment]
Buildpulse answered that they're willing to give open source discounts - but at $500/month, I doubt it's going to change anything. Thus, I'm afraid this won't be an option.

**Comment by toofar** [comment]
Ah, they got me with the $0 trick! :sob:

**Comment by toofar** [comment]
end2end tests timing out flakyly seems to be particularly bad over the last few months. We don't really get all green test runs anymore. Looking through some of the test failures in [this run](https://github.com/qutebrowser/qutebrowser/actions/runs/6874264811/job/18695618033) most of them look like the could be either the JS environment not being ready yet when the test actions ran, or, less likely, the JS environment for the tab just being borked.

Some scattered notes, too sleepy to tidy up (maybe we should split this out to a separate issue or discussion anyhow):

Always rejecting notifications (features/prompts.feature:282)
3380
clicked the button that has a JS handler
request permission
wait for log from permission callback
got nothing after
might need a timeout?

test: test_insert_mode[250-textarea.html-qute-textarea-keypress-superqutebrowser]
we insert text into the textarea
wait for onInput event listener to print text content
listener is setup via body.onload
we don't get the JS log message, no errors

tests:
  test_select_two_files_with_single_file_command
  test_select_two_files_with_multiple_files_command
  test_select_folder_for_file
click on .single_file file type input
log in onchange handler the value of the input
handler set up via body.onLoad
we don't get the log message, no errors

tests:
  :fake-key sending key to the website (features/keyinput.feature:33)
  :fake-key sending special key to the website (features/keyinput.feature:49)
log key press via keydown event handler
handler setup via window.onload
we don't get the log messages, no errors

test: Clicking on focused element (features/misc.feature:511)
click on focused element
wait for "Entering mode KeyMode.insert"
we actually get an error No element found with focus!
focus should be set by an `autofocus` attribute on the input element

I'll push a commit adding some more log messages to the JS snippets to see if anything runs. Just an idea in case I forget but we could maybe use jseval to wait for a tab to be ready (since presumably the signals aren't good enough)

**Comment by toofar** [comment]
Back to looking at tools to track failures for us. There is this pytest plugin to add annotations for failures: https://github.com/pytest-dev/pytest-github-actions-annotate-failures

But we are already getting annotations on the summary page (which is what I'm aiming for, so we can download all the annotations of the CI runs and compare them). The tests that are failing are all "wait for message" ones, and the current annotations just say the message that wasn't found, not what test failed. So I'm not sure if that plugin would be able to track what better or not. Eg

For one particular failure this is what the line in the "short summary" section says (this is perfect):

![image](https://github.com/qutebrowser/qutebrowser/assets/7419144/4a44df13-2839-41f3-a9c7-9c8af6a09641)

This is what shows up in the annotations on the CI "summary" page:

![image](https://github.com/qutebrowser/qutebrowser/assets/7419144/c7f8de29-11d2-4724-a90a-a26db3e1cac1)

This doesn't have info about the file and test that was running.

I think this is where GitHub is pulling the data for the annotation:

![image](https://github.com/qutebrowser/qutebrowser/assets/7419144/64929db5-ad1f-4728-b872-383d0a8865c7)

So possibly we could change that message to have more info in it, or change how we raise it. (Or just save a json report from pytest and attach that to the run in a separate step.)

**Comment by The-Compiler** [comment]
Ah, interesting, I wasn't aware of the pytest plugin.

The regexes we use for problem matching are here:

https://github.com/qutebrowser/qutebrowser/blob/c13089b64b103fe0e5c8526d2fbfbb5850c8c7d8/scripts/dev/ci/problemmatchers.py#L160-L187

Evidently that top one doesn't work anymore. I suspect it broke when pytest added color codes to its output there, which unfortunately need to be part of the regex. So that should be an easy fix hopefully? I'd be fine with introducing the plugin too, though.

**Comment by toofar** [comment]
Thanks for the tip! We are switching to GHA at work so maybe I'll end up copying that script.

It looks to be back to working now. We have kinda got duplicate annotations now though:

![image](https://github.com/qutebrowser/qutebrowser/assets/7419144/33b44e4f-2d36-43ed-bddb-fd53685c4a64)

With the difference that one links to the short summary at the bottom and one links to the actual failing tests.

This will only apply to new or rebased PRs but I think the currently flaky tests are pretty clear anyway.

**Comment by The-Compiler** [comment]
Took the freedom to push a couple of commits on top of your `maint/flaky_js_tests` branch:

- 14c00fcdf tox: Still accept posargs
- 9e148db5a Save screenshots when using click-element

Unfortunately the screenshots didn't actually reveal anything interesting. But when searching for `onclick` in Qt's bug tracker, I just happened to find [[QTBUG-112017] Rendering of pages without text elements is delayed](https://bugreports.qt.io/browse/QTBUG-112017) (by @toofar!), and thought I'd give the `disable-features=PaintHoldingCrossOrigin` flag there a try, given that the new severe flakiness seems to manifest on Qt 6.4 and 6.5 mostly.

Lo and behold, at least locally with `tox -e py311 -- tests/end2end/features/test_prompts_bdd.py  -k "test_javascript_confirm or test_using_contentjavascriptprompt" --count 1000 -x`, I can almost immediately reproduce a failure without that flag, but with 723a5db8f2986197e8179b4cc2143da2410b0a66, all seems well! So I pushed that to `main`, because this seems like a very big improvement without any downsides (from what I understand, the reason this is done is to avoid the dreaded white flashing on navigations, but we don't care about that during tests anyways!).

**Comment by The-Compiler** [comment]
Another tool worth looking at: https://flakytest.dev/trubleshooting/

**Comment by The-Compiler** [comment]
Codecov seems to have added something around this too! [Find failing and flaky tests with Codecov Test Analytics - Codecov](https://about.codecov.io/blog/find-failing-and-flaky-tests-with-codecov-test-analytics/)

**Comment by toofar** [comment]
Had a failure [on main](https://github.com/qutebrowser/qutebrowser/actions/runs/14434664263) (since re-ran) after merging #8537 for the bdd test "Undo the closing of a window" where it looks like the re-opened window was not active (yet).

<details>
<summary>details in screenshots</summary>

Test summary:
![Image](https://github.com/user-attachments/assets/d39940a8-e17d-47cf-b16f-fbd4f06319e3)
Actual session:
![Image](https://github.com/user-attachments/assets/0836d06a-3cc3-444f-b0af-3e4d84454c0e)
BDD scenario:
![Image](https://github.com/user-attachments/assets/0509a8e5-17b6-4522-9c17-c88c42cc7c1a)
</details>

**Comment by The-Compiler** [comment]
# `test_restart` hanging on Windows

```
tests\end2end\test_invocations.py sssss.....s.sss...s...s............... [ 13%]
Timeout (0:01:30)!
Thread 0x00000970 (most recent call first):
  File "C:\hostedtoolcache\windows\Python\3.13.2\x64\Lib\socket.py", line 295 in accept
  File "D:\a\qutebrowser\qutebrowser\.tox\py313-pyqt68\Lib\site-packages\pytest_rerunfailures.py", line 440 in run_server
  File "C:\hostedtoolcache\windows\Python\3.13.2\x64\Lib\threading.py", line 992 in run
  File "C:\hostedtoolcache\windows\Python\3.13.2\x64\Lib\threading.py", line 1041 in _bootstrap_inner
  File "C:\hostedtoolcache\windows\Python\3.13.2\x64\Lib\threading.py", line 1012 in _bootstrap

Thread 0x00000864 (most recent call first):
  File "D:\a\qutebrowser\qutebrowser\tests\end2end\test_invocations.py", line 1045 in test_restart
  File "D:\a\qutebrowser\qutebrowser\.tox\py313-pyqt68\Lib\site-packages\_pytest\python.py", line 159 in pytest_pyfunc_call
  File "D:\a\qutebrowser\qutebrowser\.tox\py313-pyqt68\Lib\site-packages\pluggy\_callers.py", line 103 in _multicall
  File "D:\a\qutebrowser\qutebrowser\.tox\py313-pyqt68\Lib\site-packages\pluggy\_manager.py", line 120 in _hookexec
  File "D:\a\qutebrowser\qutebrowser\.tox\py313-pyqt68\Lib\site-packages\pluggy\_hooks.py", line 513 in __call__
  File "D:\a\qutebrowser\qutebrowser\.tox\py313-pyqt68\Lib\site-packages\_pytest\python.py", line 1627 in runtest
  File "D:\a\qutebrowser\qutebrowser\.tox\py313-pyqt68\Lib\site-packages\_pytest\runner.py", line 174 in pytest_runtest_call
  [...]
```

that is:

https://github.com/qutebrowser/qutebrowser/blob/9447182809efe5ad41acb950b0469d93f3963be7/tests/end2end/test_invocations.py#L1040-L1045

Not sure what's going on there. Wonder if we should be using the more high-level [`psutil.Process.wait()`](https://psutil.readthedocs.io/en/latest/#psutil.Process.wait) on Windows, maybe that works better somehow.

Or we could skip the test on Windows, but with this testing `:restart` which is rather OS-specific, it would be good to keep it...

**Comment by The-Compiler** [comment]
pytest now added uploading test reports to Codecov, and it looks pretty decent:

- https://github.com/pytest-dev/pytest/pull/13778/changes
- https://app.codecov.io/gh/pytest-dev/pytest/tests
